### ISSUE TITLE
finetune代码问题。

### ISSUE THREAD

**Issue by minmie** [THUDM/ChatGLM3#184]
![image](https://github.com/THUDM/ChatGLM3/assets/40080081/c80478ed-ee40-4757-b766-5c6cdfc48b06)
请问这行代码意义是啥？为何要把label mask向右移一位，就算是当前词预测下一个词，不也是应该向左移动一位吗？

**Comment by minmie** [comment]
重新看了代码，请问向右偏移一位的目的是否是为了不计算user，assistant这些特殊token的损失？

**Comment by MGHZHANG** [comment]
我也发现了这个问题。此外再补充几个看到的关于data process以及loss mask的问题

**Comment by minmie** [comment]
@MGHZHANG 
我看了代码，说下我理解的。
1. 因为向右移动一位的操作，<|user|>，<|observation|>应该是要计算损失的，<|assistant|>则不计算损失。
2. 同1
3. 同1
4. 因为1，虽然模型没有学到eos，但是学到了<|user|>和<|observation|>，这些也是停止条件。

另外我想再请教一下，如下图，为啥是 右padding呢，生成模型不都是左padding吗
![image](https://github.com/THUDM/ChatGLM3/assets/40080081/7e037435-19c3-495c-910a-22ad76d9c1fb)

**Comment by MGHZHANG** [comment]
@minmie 
生成模型一般是在做batch inference的时候选择左padding（保证last token的logits不是无意义的padding）
实际训练的时候左padding和右padding都可以的，毕竟attention mask后pad不管在哪都不影响attention的计算。（但需要注意左pad时不要影响position id的自增）

**Comment by Btlmd** [comment]
判断 special token 是否应该被学习的标准是：**该 token 在推理时由推理系统加入，不学习；由模型生成，学习**。在当前的实现中，一个 take loss 的 assistant 回复后的下一个 token，无论是何种 special token（目前可能是 `<|assistant|>`, `<|user|>` 或 `<|observation|>`）都会 take loss。这里就不妨用 sanity check 的输出作为例子说明

```bash
Sanity Check >>>>>>>>>>>>>
           '[gMASK]':  64790 ->   -100
               'sop':  64792 ->   -100
        '<|system|>':  64794 ->   -100 # <|system|> 由推理系统加入，没有 take loss
                  '':  30910 ->   -100
                '\n':     13 ->   -100
            'Answer':  20115 ->   -100
               'the':    267 ->   -100
         'following':   1762 ->   -100
         'questions':   2554 ->   -100
                'as':    362 ->   -100
              'best':   1077 ->   -100
                'as':    362 ->   -100
               'you':    344 ->   -100
               'can':    457 ->   -100
                 '.':  30930 ->   -100
               'You':    809 ->   -100
              'have':    431 ->   -100
            'access':   1675 ->   -100
                'to':    289 ->   -100
               'the':    267 ->   -100
         'following':   1762 ->   -100
             'tools':   4159 ->   -100
                 ':':  30954 ->   -100
                '\n':     13 ->   -100
                '["':   4812 ->   -100
                    ...
                '"]':   5515 ->   -100
          '<|user|>':  64795 ->   -100 # <|user|> 由系统加入，没有 take loss
                  '':  30910 ->   -100
                '\n':     13 ->   -100
                 'I':    307 ->   -100
                 "'":  30953 ->   -100
                 'm':  30924 ->   -100
             'troub':   9793 ->   -100
              'lesh':  10761 ->   -100
            'ooting':  16954 ->   -100
              'some':    635 ->   -100
          'requests':   8563 ->   -100
                    ...
               'Let':   2641 ->   -100
                'me':    496 ->   -100
              'know':    683 ->   -100
               'the':    267 ->   -100
          'response':   3010 ->   -100
           'details':   3296 ->   -100
                 '.':  30930 ->   -100
     '<|assistant|>':  64796 ->   -100 # 第一个 <|assistant|> 由系统加入，没有 take loss
                  '':  30910 ->  30910 # 从这里开始的部分是需要模型生成的，take loss
                '\n':     13 ->     13
                 'I':    307 ->    307
              'need':    720 ->    720
                'to':    289 ->    289
               'use':    792 ->    792
               'the':    267 ->    267
              'send':   3862 ->   3862
              'Http':  14428 ->  14428
           'Request':   7936 ->   7936
              'tool':   2583 ->   2583
                'to':    289 ->    289
              'send':   3862 ->   3862
               'the':    267 ->    267
              'POST':  24838 ->  24838
           'request':   2555 ->   2555
                 '.':  30930 ->  30930
     '<|assistant|>':  64796 ->  64796 # 第二个 <|assistant|> 由模型生成，take loss
              'send':   3862 ->   3862
              'Http':  14428 ->  14428
           'Request':   7936 ->   7936
                '\n':     13 ->     13
               '```':  16014 ->  16014
            'python':  23720 ->  23720
                '\n':     13 ->     13
              'tool':  22268 ->  22268
                 '_':  30962 ->  30962
              'call':  10372 ->  10372
                 '(':  30946 ->  30946
                    ...
                '})':   4722 ->   4722
                '\n':     13 ->     13
                '``':  10846 ->  10846
                 '`':  31040 ->  31040
   '<|observation|>':  64797 ->  64797 # <|observation|> 由模型生成，take loss
                  '':  30910 ->   -100 # 从这里开始的部分是推理系统加入的，没有 take loss
                '\n':     13 ->   -100
            'Status':  16297 ->   -100
              'Code':   7025 ->   -100
                 ':':  30954 ->   -100
                  '':  30910 ->   -100
                 '2':  30943 ->   -100
                 '0':  30940 ->   -100
                 '0':  30940 ->   -100
                 '.':  30930 ->   -100
                    ... 
                '}}':   2059 ->   -100
     '<|assistant|>':  64796 ->   -100 # 第三个 <|assistant|> 由系统加入，没有 take loss
                  '':  30910 ->  30910
                '\n':     13 ->     13
               'The':    353 ->    353
           'request':   2555 ->   2555
               'was':    401 ->    401
        'successful':   3777 ->   3777
                    ...
                 '.':  30930 ->  30930
               'The':    353 ->    353
          'response':   3010 ->   3010
            'status':   4123 ->   4123
              'code':   2887 ->   2887
               'was':    401 ->    401
                  '':  30910 ->  30910
                 '2':  30943 ->  30943
                 '0':  30940 ->  30940
                 '0':  30940 ->  30940
                 '.':  30930 ->  30930
                  '':      2 ->      2  # <eos> 由模型生成，take loss；此处若为 <|user|> ，也会 take loss
                    ...
<<<<<<<<<<<<< Sanity Check
```

**Comment by Btlmd** [comment]
> ![image](https://user-images.githubusercontent.com/40080081/280617906-c80478ed-ee40-4757-b766-5c6cdfc48b06.png) 请问这行代码意义是啥？为何要把label mask向右移一位，就算是当前词预测下一个词，不也是应该向左移动一位吗？

在 `format_conversation` 函数中，loss_mask 的含义是，某个 token 输入后，所预测的 token 是否需要 take loss；而输入模型的 target 的含义是，这个 token 相同位置应该预测什么，因此需要右移一位。

例如，在 `format_conversation` 函数中的 loss mask 为 

```bash
<|user|>       -> 0 
hello          -> 0 
<|assistant|>  -> 1
hi             -> 1
<|user|>       -> 0
```

对应 `target_based_loss_mask` 就应为

```bash
<|user|>       -> 0 
hello          -> 0 
<|assistant|>  -> 0
hi             -> 1
<|user|>       -> 1
```
实际模型应该学习的也是 `hi` 和 `<|user|>`

> 重新看了代码，请问向右偏移一位的目的是否是为了不计算user，assistant这些特殊token的损失？

我觉得这样理解也可以。

**Comment by Btlmd** [comment]
> 我也发现了这个问题。此外再补充几个看到的关于data process以及loss mask的问题
> 
> 1. 多轮格式下，特殊role id的predict loss是否计算？从代码上来看，<|user|>、<|system|>、<|observation|>的predict loss不计算，但是<|assistant|>的loss会计算。
> 2. 我感觉<|observation|>的loss还是需要计算的，首先是我看chat方法里在generate时设置了<|observation|>作为其中之一的结束token，且发现模型在调用tool生成时（生成完tool name和parameters后）一般也会生成<|observation|>，所以模型是可以学到下一个token是<|observation|>的，推测训练的时候应该没有mask掉<|observation|>的loss。
> 3. chat方法在生成前会在先手动拼上一个<|assistant|>后再generate，所以训练时对于轮次开始的<|assistant|>是否也可以mask掉不用模型学习
> 4. 在多轮格式下，eos token id只在最后一轮结束后才拼上，也就是说中间的assistant轮次是没有eos id的，这样训练是否会造成模型对中间轮次的回复没有学到怎么结束生成

1-3. 见前文对 *判断 special token 是否应该被学习的标准*
4. 对于中间轮次的对话，模型学到的是输出 `<|user|>` 或 `<|observation|>` 作为终止 / 暂停生成的条件

**Comment by Btlmd** [comment]
> 另外我想再请教一下，如下图，为啥是 右padding呢，生成模型不都是左padding吗 ![image](https://user-images.githubusercontent.com/40080081/280899261-7e037435-19c3-495c-910a-22ad76d9c1fb.png)

- 对于 RoPE，训练时左右 padding 并不影响训练效果，只要 position id 是连续的就可以
- 推理时左 padding 也是为了实现方便（如果 attention mask 和 position id 配合得当，也是可以右 padding 的）

**Comment by treya-lin** [comment]
> > 我也发现了这个问题。此外再补充几个看到的关于data process以及loss mask的问题
> > 
> > 1. 多轮格式下，特殊role id的predict loss是否计算？从代码上来看，<|user|>、<|system|>、<|observation|>的predict loss不计算，但是<|assistant|>的loss会计算。
> > 2. 我感觉<|observation|>的loss还是需要计算的，首先是我看chat方法里在generate时设置了<|observation|>作为其中之一的结束token，且发现模型在调用tool生成时（生成完tool name和parameters后）一般也会生成<|observation|>，所以模型是可以学到下一个token是<|observation|>的，推测训练的时候应该没有mask掉<|observation|>的loss。
> > 3. chat方法在生成前会在先手动拼上一个<|assistant|>后再generate，所以训练时对于轮次开始的<|assistant|>是否也可以mask掉不用模型学习
> > 4. 在多轮格式下，eos token id只在最后一轮结束后才拼上，也就是说中间的assistant轮次是没有eos id的，这样训练是否会造成模型对中间轮次的回复没有学到怎么结束生成
> 
> 1-3. 见前文对 _判断 special token 是否应该被学习的标准_ 4. 对于中间轮次的对话，模型学到的是输出 `<|user|>` 或 `<|observation|>` 作为终止 / 暂停生成的条件

你好，这里说对于中间轮次的对话，模型学到的输出是`<|user|>` 或 `<|observation|>` 作为终止 / 暂停生成的条件,但是我用多轮微调以后，它并没有以次作为暂停生成的条件。可以看看是为什么吗？https://github.com/THUDM/ChatGLM3/issues/980

![image](https://github.com/THUDM/ChatGLM3/assets/86940562/c6fc2a2d-48d9-4ccd-844e-9165623526be)

**Comment by STHSF** [comment]
![image](https://github.com/THUDM/ChatGLM3/assets/8037442/796307dc-eec1-4cf5-aa30-1bc29f244a1d)
有个问题想请教一下，我在使用chatglm3 的tokenizer的时候对比中英文状态下的token发现会有多余的空字符出现，请问这个正常吗？为啥会出现空字符呢？
